numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1434 @@
+};
+
+
+/**
+ * Adds known sequence types to the global type dictionary, note that when
+ * a DType is passed in, this lookup may be ignored.
+ *
+ * @return -1 on error 0 on success
+ */
+static int
+_prime_global_pytype_to_type_dict()
+{
+    int res;
+
+    /* Add the basic Python sequence types */
+    res = PyDict_SetItem(_global_pytype_to_type_dict,
+                         (PyObject *)&PyList_Type, Py_None);
+    if (res < 0) {
+        return -1;
+    }
+    res = PyDict_SetItem(_global_pytype_to_type_dict,
+                         (PyObject *)&PyTuple_Type, Py_None);
+    if (res < 0) {
+        return -1;
+    }
+    /* NumPy Arrays are not handled as scalars */
+    res = PyDict_SetItem(_global_pytype_to_type_dict,
+                         (PyObject *)&PyArray_Type, Py_None);
+    if (res < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+
+/**
+ * Add a new mapping from a python type to the DType class.
+ *
+ * This assumes that the DType class is guaranteed to hold on the
+ * python type (this assumption is guaranteed).
+ * This functionality supercedes ``_typenum_fromtypeobj``.
+ *
+ * @param DType DType to map the python type to
+ * @param pytype Python type to map from
+ * @param userdef Whether or not it is user defined. We ensure that user
+ *        defined scalars subclass from our scalars (for now).
+ */
+NPY_NO_EXPORT int
+_PyArray_MapPyTypeToDType(
+        PyArray_DTypeMeta *DType, PyTypeObject *pytype, npy_bool userdef)
+{
+    PyObject *Dtype_obj = (PyObject *)DType;
+
+    if (userdef) {
+        /*
+         * It seems we did not strictly enforce this in the legacy dtype
+         * API, but assume that it is always true. Further, this could be
+         * relaxed in the future. In particular we should have a new
+         * superclass of ``np.generic`` in order to note enforce the array
+         * scalar behaviour.
+         */
+        if (!PyObject_IsSubclass((PyObject *)pytype, (PyObject *)&PyGenericArrType_Type)) {
+            PyErr_Format(PyExc_RuntimeError,
+                    "currently it is only possible to register a DType "
+                    "for scalars deriving from `np.generic`, got '%S'.",
+                    (PyObject *)pytype);
+            return -1;
+        }
+    }
+
+    /* Create the global dictionary if it does not exist */
+    if (NPY_UNLIKELY(_global_pytype_to_type_dict == NULL)) {
+        _global_pytype_to_type_dict = PyDict_New();
+        if (_global_pytype_to_type_dict == NULL) {
+            return -1;
+        }
+        if (_prime_global_pytype_to_type_dict() < 0) {
+            return -1;
+        }
+    }
+
+    int res = PyDict_Contains(_global_pytype_to_type_dict, (PyObject *)pytype);
+    if (res < 0) {
+        return -1;
+    }
+    else if (res) {
+        PyErr_SetString(PyExc_RuntimeError,
+                "Can only map one python type to DType.");
+        return -1;
+    }
+
+    return PyDict_SetItem(_global_pytype_to_type_dict,
```

Comment:
Will this hold onto references forever? What about type factories and so on, where this could get arbitrarily large?

Reply:
Yes, should I expand the comment at the top of the function?

Obviously that is not ideal, but you _could_ do a weak mapping eventually. Unfortunately, I think you may also need circular reference resolution to really support thing (or possibly tie the live-span to the python type using a weakref, although then you need to do it even when the DType is not part of this mapping).

Lets be clear, there is a natural ref-cycle between type<->DType I have mentioned often I feel. And if someone has a genius solution to it, I am all ears. But, I am not smart enough to resolve it directly.
Both DType and type must keep each-other alive (usually) after all, the only way to try and resolve it would be cramming the dtype into the type itself (requiring to modify the type) and making dtypes without a clear type possibly weird (you don't have to be part of this mapping).  Plus it would require that the user creates two classes instead of one, because generally, you still need to extend the dtype storage anyway.

Reply:
To be clear, there may not be a good solution, and it boils down to *how* we add things to this mapping in the end.  The mapping can be weak (in most cases), but to resolve dynamically created `pytype<->DType` pairs, you must use cyclic garbage collection.
So when you add something to the mapping for a dynamically created pytype, things are different. And we could force the user to attach a reference of their `DType` to the `type` in the process (however exactly) to allow adding it to this map (or generally lookup).  That is clean *for dyncamically created pytypes*.  For most heaptypes, we could probably just do that even, since we can inject `__associated_numpy_DType__` into their dict, thus automatically enabling cyclic reference counting.  And maybe that is already a nice solution, just use cyclic refcounting and then allow it to resolve `DType<->pytype` pairs (if their are heaptypes).  I guess it may be a user has `__slots__` or so to make their type read-only, that would be their problem then, its a *much* simpler update then having to create a pair of classes to begin with IMO.  I honestly feel the mental load of having a single Pytype+DType is pretty high, but maybe there is a nice way to formalize it that makes it all clear in a whizz.

Everything else would require the DType to live on the pytype to begin with (always), there could be no `DType` class as such, it would need to refer for everything to the pytype through lookup on the instance or so).  That seems complicated to me, but maybe I am missing some great pattern here.  The trick would be that the DType class is light-weight, and does not know about the pytype, only its instances do.

Reply:
The solution to break refcount cycles is to make one of the links in the cycle a weakref, as suggested. I'm okay with cycles, just not okay with a global dict that lives forever. The issue is very practical... Suppose one writes the following code with `a = 1 meter` (and assuming each new unit is a new dtype)
```
b = 1
for i in range(500):
    b *= a
```

In this case you have 500 new dtypes being stuck in this dict forever, 499 of them useless.

Reply:
I will add a paragraph on this to NEP 42, with the dunder method solution.  The mapping will have to be complicated, but that is a technical program that can be adjourned for a very long time.  Hell, I quite frankly think it is fine to release with this as a "known bug" (but fixable) that dynamically created `pytype<->DType` *pairs* are immortal.

Reply:
@hameerabbasi no, it is *not* that simple, and this type of unit is *not* an issue or relevant here unless you also create `meter_scalar(3)` and want `np.array(meter_scalar(3))` to automatically go to your new dtype.  That is *not* how the NEP describes you should be doing it, there are parametric dtypes for a reason.

It is far less practically important than you think, I frankly believe.  Each new unit is a new *dtype instance*, making each new unit a new DType is very impractical on many levels.  I.e. ufunc loop lookup/casting would have to be defined for some kind of super-class and than figure out how it works for that subclass?

There are imaginable use-cases. E.g. an `Enum` DType for `Enum` types (which are metaclasses). But, quite honestly, I think `unit` is a contrived one (show me an example of someone putting the unit on the *class* and not on the instance).  And except `Enum` I cannot think of any actual use-case (Categorical does not count, because its polymorphic(?) enum, there is no actual type to it).

So, yes its important, but its not *central* to any serious use-case I am aware of, such as Units, DNA, special integers, masked integers, datetimes, ...

Reply:
Ah, if the instance is not cached then it’s fine I guess. Parametrised dtypes are good enough.

Reply:
Yeah, I *really* want it to be resolvable at least in theory in the future though, so thanks for prodding me towards the "just add a special attribute" to the python type when/if registering to be discoverable like this.  It seems like a reasonable solution to me, and simple to understand for an implementor:  They get an error "making `np.array(scalar)` without dtype work, requires numpy to add `__associated_array_dtype__` to the type, this failed".  Would be good to enforce it from the start when we have public API (even if we may not fully support it immediately).

Reply:
It would also be good to track this in an issue when this PR gets merged.